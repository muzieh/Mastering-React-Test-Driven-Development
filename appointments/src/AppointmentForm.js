import React, { useState } from 'react';

export const AppointmentForm = ({
  selectableServices,
  service,
  onSubmit,
}) => {
  const [appointment, setAppointment] = useState({ service });
  return (
    <form
      id="appointment"
      onSubmit={() => onSubmit(appointment)} >
      <label htmlFor="service">Service</label>
      <select id="service" name="service" value={service} readOnly>
        <option />
        {selectableServices.map((s) => (
          <option key={s} value={s}>
            {s}
          </option>
        ))}
        );
      </select>
    </form>
  );
};

AppointmentForm.defaultProps = {
  selectableServices: [
    'Cut',
    'Blow-dry',
    'Cut & color',
    'Beard trim',
    'Cut & beard trim',
    'Extensions',
  ],
};
