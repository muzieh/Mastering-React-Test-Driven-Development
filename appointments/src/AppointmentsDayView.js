import React, {useState} from 'react';

const appointmentTimeOfDay = (startsAt) => {
  const [h, m] = new Date(startsAt).toTimeString().split(':');
  return `${h}:${m}`;
};

export const Appointment = ({customer, startsAt, service, note, stylist}) => (
  <div>
    <h2>{appointmentTimeOfDay(startsAt)}</h2>
    <table>
      <tbody>
      <tr>
        <td>{customer.firstName}</td>
        <td>{customer.lastName}</td>
        <td>{stylist}</td>
        <td>{service}</td>
        <td>{note}</td>
      </tr>
      </tbody>
    </table>
  </div>
);

export const AppointmentsDayView = ({ appointments }) => {
  const [selectedAppointment, setSelectedAppointment] = useState(
    0
  );

  return (
    <div id="appointmentsDayView">
      <ol>
        {appointments.map((appointment, i) => (
          <li key={appointment.startsAt}>
            <button
              type="button"
              onClick={() => setSelectedAppointment(i)}>
              {appointmentTimeOfDay(appointment.startsAt)}
            </button>
          </li>
        ))}
      </ol>
      {appointments.length === 0 ? (
        <p>There are no appointments scheduled for today.</p>
      ) : (
        <Appointment {...appointments[selectedAppointment]} />
      )}
    </div>
  );
};
