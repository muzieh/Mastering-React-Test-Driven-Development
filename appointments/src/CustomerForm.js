import React from 'react';

export const CustomerForm = (props) => {
  return <form id='customer'>
    <input type="text" name="firstName"/>
  </form>;
};
