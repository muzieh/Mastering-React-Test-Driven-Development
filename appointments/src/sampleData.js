import {lorem, name, phone} from 'faker';

const today = new Date();

Array.prototype.unique = function () {
  return this.filter(function (value, index, self) {
    return self.indexOf(value) === index;
  });
};

Array.prototype.pickRandom = function () {
  return this[Math.floor(Math.random() * this.length)];
};

const at = (hours) => today.setHours(hours, 0);
const createFakeCustomer = () => ({
  firstName: name.firstName(),
  lastName: name.lastName(),
  phoneNumber: phone.phoneNumber(),
});

const services = ['ser1', 'ser2', 'ser3'];

const stylists = [1, 2, 3, 4, 5, 6]
  .map(() => name.firstName())
  .unique();

const createFakeAppointment = () => ({
  customer: {...createFakeCustomer()},
  service: services.pickRandom(),
  stylist: stylists.pickRandom(),
  notes: lorem.paragraph(),
});

console.log(createFakeCustomer());

export const sampleAppointments = [
  {startsAt: at(8), ...createFakeAppointment()},
  {startsAt: at(9), ...createFakeAppointment()},
  {startsAt: at(10), ...createFakeAppointment()},
  {startsAt: at(11), ...createFakeAppointment()},
  {startsAt: at(12), ...createFakeAppointment()},
  {startsAt: at(13), ...createFakeAppointment()},
  {startsAt: at(14), ...createFakeAppointment()},
  {startsAt: at(15), ...createFakeAppointment()},
  {startsAt: at(16), ...createFakeAppointment()},
  {startsAt: at(17), ...createFakeAppointment()},
];
