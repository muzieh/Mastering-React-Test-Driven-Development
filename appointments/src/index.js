import React from 'react';
import ReactDOM from 'react-dom';
import { AppointmentsDayView } from './AppointmentsDayView';
import { sampleAppointments } from './sampleData';
import { CustomerForm } from './CustomerForm';


ReactDOM.render(
  <div>
    <AppointmentsDayView appointments={sampleAppointments} />
    <CustomerForm />
  </div>,
  document.getElementById('root')
);
