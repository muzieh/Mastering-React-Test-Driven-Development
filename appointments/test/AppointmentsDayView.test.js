import React from 'react';
import ReactDOM from 'react-dom';
import TestRenderer from 'react-test-renderer';
import ReactTestUtils from 'react-dom/test-utils';
import {Appointment, AppointmentsDayView,} from '../src/AppointmentsDayView';

describe('Appointment', () => {
  let container;
  let customer;
  let appointment;

  beforeEach(() => {
    container = document.createElement('div');

    customer = {
      firstName: 'Ashley',
      lastName: 'Bings',
    };

    appointment = {
      startsAt: new Date().setHours(13, 20),
      stylist: 'Ben Shapiro',
      service: 'perm',
      note: 'some note',
      customer: customer,
    };
  });

  const render = (component) =>
    ReactDOM.render(component, container);

  it('renders content in table', () => {
    render(<Appointment customer={customer}/>);
    expect(container.querySelector('table')).not.toBeNull();
  });

  it('renders appointment header', () => {
    render(<Appointment customer={customer}/>);
    expect(container.querySelector('h2')).not.toBeNull();
  });

  it('renders appointment start time in header', () => {
    render(<Appointment {...appointment} />);
    expect(container.querySelector('h2').textContent).toMatch(
      '13:20'
    );
  });

  it('renders the customer first name', () => {
    render(<Appointment customer={customer}/>);
    expect(container.textContent).toMatch('Ashley');
  });

  it('renders the customer last name', () => {
    render(<Appointment customer={customer}/>);
    expect(container.textContent).toMatch('Bings');
  });

  it('renders the customer stylist', () => {
    render(<Appointment {...appointment} />);
    expect(container.textContent).toMatch('Ben Shapiro');
  });

  it('renders the customer service', () => {
    render(
      <Appointment
        customer={appointment.customer}
        service={appointment.service}
      />
    );
    expect(container.textContent).toMatch('perm');
  });

  it('renders the customer note', () => {
    render(<Appointment {...appointment} />);
    expect(container.textContent).toMatch('some note');
  });

  it('renders another customer first name', () => {
    customer = {firstName: 'Jordan'};
    render(<Appointment customer={customer}/>);
    expect(container.textContent).toMatch('Jordan');
  });
});

describe('AppointmentsDayView', () => {
  const today = new Date();
  const appointments = [
    {
      startsAt: today.setHours(12, 0),
      customer: {firstName: 'Ashley'},
      note: 'some note',
      service: 'perm',
      stylist: 'Ben Shapiro',
    },
    {
      startsAt: today.setHours(13, 0),
      customer: {firstName: 'Jordan'},
    },
  ];
  let container;

  beforeEach(() => {
    container = document.createElement('div');
  });

  const render = (component) =>
    ReactDOM.render(component, container);

  it('renders a div with the right id', () => {
    render(<AppointmentsDayView appointments={[]} />);
    expect(
      container.querySelector('div#appointmentsDayView')
    ).not.toBeNull();
  });

  it('renders multiple appointments in an ol element', () => {
    render(<AppointmentsDayView appointments={appointments} />);
    expect(container.querySelector('ol')).not.toBeNull();
    expect(container.querySelector('ol').children).toHaveLength(2);
  });

  it('renders each appointment in an li', () => {
    render(<AppointmentsDayView appointments={appointments} />);
    expect(container.querySelectorAll('li')).toHaveLength(2);
    expect(
      container.querySelectorAll('li')[0].textContent
    ).toEqual('12:00');
    expect(
      container.querySelectorAll('li')[1].textContent
    ).toEqual('13:00');
  });

  it('initially shows a message saying there are no appointments today', () => {
    render(<AppointmentsDayView appointments={[]} />);
    expect(container.textContent).toMatch(
      'There are no appointments scheduled for today.'
    );
  });

  it('selects the first appointment by default', () => {
    render(<AppointmentsDayView appointments={appointments} />);
    expect(container.textContent).toMatch('Ashley');
  });

  it('has a button element in each li', () => {
    render(<AppointmentsDayView appointments={appointments} />);
    expect(container.querySelectorAll('li > button')).toHaveLength(
      2
    );
    expect(
      container.querySelectorAll('li > button')[0].type
    ).toEqual('button');
  });

  it('renders another appointment when selected', () => {
    render(<AppointmentsDayView appointments={appointments}/>);
    const button = container.querySelectorAll('button')[1];
    ReactTestUtils.Simulate.click(button);
    expect(container.textContent).toMatch('Jordan');
  });

  it('snapshot test', () => {
    const tree = TestRenderer.create(
      <AppointmentsDayView appointments={appointments}/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
