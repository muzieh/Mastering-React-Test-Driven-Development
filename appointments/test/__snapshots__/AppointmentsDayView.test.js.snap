// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppointmentsDayView snapshot test 1`] = `
<div
  id="appointmentsDayView"
>
  <ol>
    <li>
      <button
        onClick={[Function]}
        type="button"
      >
        12:00
      </button>
    </li>
    <li>
      <button
        onClick={[Function]}
        type="button"
      >
        13:00
      </button>
    </li>
  </ol>
  <div>
    <h2>
      12:00
    </h2>
    <table>
      <tbody>
        <tr>
          <td>
            Ashley
          </td>
          <td />
          <td>
            Ben Shapiro
          </td>
          <td>
            perm
          </td>
          <td>
            some note
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
